### PR TITLE
Fix isolation analysis for start actions

### LINKED ISSUE #41691: [Bug]: Invalid inferring of `isolated` with start actions when the arguments are not isolated expressions
### Description

$title.

### Steps to Reproduce

```ballerina
import ballerina/http;

type UserRegistration record {
    string name;
    string email;
};

service on new http:Listener(8080) {
    resource function get h1(UserRegistration user) {
        _ = start sendEmailNotification(user);
    }
}

isolated function sendEmailNotification(UserRegistration user) returns error? {
}
```

Here, it seems like the resource method is inferred as `isolated` - no warning - despite the argument not being an isolated expression.

But, if we explicitly mark the resource method as `isolated`, it results in a compilation error for the argument.

```ballerina
import ballerina/http;

type UserRegistration record {
    string name;
    string email;
};

service on new http:Listener(8080) {
    isolated resource function get h1(UserRegistration user) {
        _ = start sendEmailNotification(user);
    }
}

isolated function sendEmailNotification(UserRegistration user) returns error? {
}
```

```cmd
$ bal run --offline sample.bal 
Compiling source
        sample.bal
ERROR [sample.bal:(10:41,10:45)] invalid start action accessing a non isolated expression in an argument of a async call of an 'isolated' function
```

### Affected Version(s)

2201.8.0

### OS, DB, other environment details and versions

_No response_

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by LakshanWeerasinghe**
## Purpose
$subject

FIxes #41691 

Related PR #41772 

## Remarks
This PR includes only fixing isolation analysis with start actions. The compiler crashes happens when passing mapping and list constructors will be fixed by #41772.

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-unit-test/src/test/resources/test-src/isolation-analysis/isolation_warnings_for_service_methods.bal
[the remark this answers is not in the document]
Comment (LakshanWeerasinghe) [reply tests/jballerina-unit-test/src/test/resources/test-src/isolation-analysis/isolation_warnings_for_service_methods.bal:1]: added in [commit](https://github.com/ballerina-platform/ballerina-lang/pull/42366/commits/a63eab90078fda6d54817fd5f77abecd890a9963).

**Comment by MaryamZi** [comment]
> We don't seem to be logging a warning for
> 
> ```ballerina
> import ballerina/http;
> 
> type UserRegistration record {
>     string name;
>     string email;
> };
> 
> class Obj {
>     int x = 1;
> 
>     function fn() {
> 
>     }
> }
> 
> function getObj(Obj obj) returns Obj => obj;
> 
> service on new http:Listener(8080) {
>     resource function get h1(UserRegistration user) {
>         Obj obj = new;
>         _ = start sendEmailNotification(getObj(obj));
>     }
> }
> 
> isolated function sendEmailNotification(object {} ob) returns error? {
> }
> ```

This isn't fixed yet?

**Comment by LakshanWeerasinghe** [comment]
> Fix issue with infering func isolation

Fixed in [commit](https://github.com/ballerina-platform/ballerina-lang/pull/42366/commits/e05696da3cb14fe9778e0d25435af4eff4b6c1db).

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java [hunk 1]
```diff
@@ -1273,6 +1274,10 @@ public void visit(BLangSimpleVarRef varRefExpr) {
 
         if (!recordFieldDefaultValue && !objectFieldDefaultValueRequiringIsolation && enclInvokable != null &&
                 isReferenceToVarDefinedInSameInvokable(symbol.owner, enclInvokable.symbol)) {
+            if (this.inIsolatedStartAction
+                    && !isSubtypeOfReadOnlyOrIsolatedObjectOrInferableObject(symbol.owner, symbol.getType())) {
+                inferredIsolated = false;
+            }
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java:1280]: Can't this happen outside the outer check?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java [hunk 2]
```diff
@@ -4125,7 +4133,8 @@ private boolean inferFunctionIsolation(BSymbol symbol, IsolationInferenceInfo fu
         }
 
         for (BLangExpression dependsOnArg : functionIsolationInferenceInfo.dependsOnFuncCallArgExprs) {
-            if (!isIsolatedExpression(dependsOnArg)) {
+            if (!isIsolatedExpression(dependsOnArg, false, false, new ArrayList<>(), true,
+                    publiclyExposedObjectTypes, classDefinitions, moduleLevelVariables, new HashSet<>())) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java:4137]: Shouldn't we pass `unresolvedSymbols` here? Shall we add a test too?

File: tests/jballerina-unit-test/src/test/resources/test-src/isolation-analysis/isolation_analysis_for_async_actions.bal [hunk 3]
```diff
@@ -0,0 +1,36 @@
+service / on new Listener() {
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/isolation-analysis/isolation_analysis_for_async_actions.bal:1]: Add the licence header.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java (hunk 1 above)
Comment (LakshanWeerasinghe) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java:1280]: At the outside of the outer check we analyse symbols with `recordFieldDefaultValue` and `objectFieldDefaultValueRequiringIsolation` being true. For those I don't think we need to check its a subtype of readonly or isolated object.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java (hunk 2 above)
Comment (LakshanWeerasinghe) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java]: The existing tests will cover this scenario.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java (hunk 1 above)
Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java:1280]: But for those `this.inIsolatedStartAction` won't be true, right?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java (hunk 2 above)
Comment (LakshanWeerasinghe) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java]: Updated in [commit](https://github.com/ballerina-platform/ballerina-lang/pull/42366/commits/aef9b040ac3a3a01bc0d9108cd0ebaf5bf0405db).

Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java]: But that should have resulted in a stack overflow-ish error, right? If we are recursively checking the same isolated-ness of the same construct?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java (hunk 1 above)
Comment (LakshanWeerasinghe) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java:1280]: Yes. But outer was there to handle both `true` and `false` scenarios. We have add this new check to validate start actions specifically. Which wasn't being analysed previously.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java (hunk 2 above)
Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java]: Example:

```ballerina
function fn(any x = ()) returns any {
    _ = start fn(fn());
}
```

Comment (LakshanWeerasinghe) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java]: Thanks!
When we don't pass unresolved symbols, for the provided sample it gives SO.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java (hunk 1 above)
Comment (LakshanWeerasinghe) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java:1297]: Then we need `isReferenceToVarDefinedInSameInvokable` check as well.

Comment (LakshanWeerasinghe) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java]: When I am running the test locally previously seems I haven't added the `return` statement. We can move this to the outside of the outer if. Updated in [commit](https://github.com/ballerina-platform/ballerina-lang/pull/42366/commits/a22e04b5a5c73c4d6a0327088fd0126d7125d7a5).

**Review by MaryamZi — APPROVED** [review]

**Pull request merged**